File: web/Back/Integration.Elise.Services/Services/DashboardService.cs
/**
 * @file     DashboardService.cs
 * @module   NeoLeadge — Deployment Manager
 * @author   [dev]
 * @date     2026-03-27
 * @desc     Dashboard statistics and metrics service
 */

using Integration.Elise.Services.Infrastructure;
using Integration.Elise.Services.Interfaces;
using Integration.Elise.Services.Models;
using Integration.Elise.Services.Models.DTOs;
using Integration.Elise.Services.Models.Enums;
using Microsoft.EntityFrameworkCore;
using Serilog;

namespace Integration.Elise.Services.Services;

/// <inheritdoc />
public class DashboardService : IDashboardService
{
    private readonly ApplicationDbContext _db;

    public DashboardService(ApplicationDbContext db)
    {
        _db = db;
    }

    /// <inheritdoc />
    public async Task<Result<DashboardStatsDto>> GetDashboardStatsAsync(CancellationToken ct = default)
    {
        var totalProjects = await _db.Projects.CountAsync(ct);
        var activeProjects = await _db.Projects
            .Where(p => p.Status != ProjectStatus.Completed && p.Status != ProjectStatus.Archived)
            .CountAsync(ct);
        var completedProjects = await _db.Projects
            .Where(p => p.Status == ProjectStatus.Completed)
            .CountAsync(ct);
        var archivedProjects = await _db.Projects
            .Where(p => p.Status == ProjectStatus.Archived)
            .CountAsync(ct);

        var now = DateTime.UtcNow;
        var overdueProjects = await _db.Projects
            .Where(p => p.EndDate < now && p.Status != ProjectStatus.Completed && p.Status != ProjectStatus.Archived)
            .CountAsync(ct);

        var startOfMonth = new DateTime(now.Year, now.Month, 1, 0, 0, 0, DateTimeKind.Utc);
        var projectsThisMonth = await _db.Projects
            .Where(p => p.CreatedAt >= startOfMonth)
            .CountAsync(ct);

        var projectsByStatus = await _db.Projects
            .GroupBy(p => p.Status.ToString())
            .Select(g => new { Status = g.Key, Count = g.Count() })
            .ToDictionaryAsync(x => x.Status, x => x.Count, ct);

        var projectsByPriority = await _db.Projects
            .GroupBy(p => p.Priority.ToString())
            .Select(g => new { Priority = g.Key, Count = g.Count() })
            .ToDictionaryAsync(x => x.Priority, x => x.Count, ct);

        var workloads = await GetProjectManagerWorkloadsAsync(ct);
        var recentActivities = await GetRecentActivityAsync(10, ct);

        var stats = new DashboardStatsDto(
            TotalProjects: totalProjects,
            ActiveProjects: activeProjects,
            CompletedProjects: completedProjects,
            ArchivedProjects: archivedProjects,
            OverdueProjects: overdueProjects,
            ProjectsThisMonth: projectsThisMonth,
            ProjectsByStatus: projectsByStatus,
            ProjectsByPriority: projectsByPriority,
            ProjectManagerWorkloads: workloads.Value ?? new List<ProjectManagerWorkloadDto>(),
            RecentActivities: recentActivities.Value ?? new List<RecentActivityDto>()
        );

        return Result<DashboardStatsDto>.Ok(stats);
    }

    /// <inheritdoc />
    public async Task<Result<Dictionary<string, int>>> GetProjectsByStatusAsync(CancellationToken ct = default)
    {
        var result = await _db.Projects
            .GroupBy(p => p.Status.ToString())
            .Select(g => new { Status = g.Key, Count = g.Count() })
            .ToDictionaryAsync(x => x.Status, x => x.Count, ct);

        return Result<Dictionary<string, int>>.Ok(result);
    }

    /// <inheritdoc />
    public async Task<Result<List<ProjectManagerWorkloadDto>>> GetProjectManagerWorkloadsAsync(CancellationToken ct = default)
    {
        var managers = await _db.AppUsers
            .Where(u => u.Role == UserRole.ProjectManager && u.IsActive)
            .ToListAsync(ct);

        var workloads = new List<ProjectManagerWorkloadDto>();

        foreach (var manager in managers)
        {
            var projects = await _db.Projects
                .Where(p => p.ProjectManagerId == manager.Id)
                .ToListAsync(ct);

            var now = DateTime.UtcNow;
            workloads.Add(new ProjectManagerWorkloadDto(
                ManagerId: manager.Id,
                ManagerName: $"{manager.FirstName} {manager.LastName}",
                ManagerEmail: manager.Email,
                TotalProjects: projects.Count,
                InProgressProjects: projects.Count(p => p.Status == ProjectStatus.InProgress),
                CompletedProjects: projects.Count(p => p.Status == ProjectStatus.Completed),
                OverdueProjects: projects.Count(p => p.EndDate < now && p.Status != ProjectStatus.Completed && p.Status != ProjectStatus.Archived)
            ));
        }

        return Result<List<ProjectManagerWorkloadDto>>.Ok(workloads);
    }

    /// <inheritdoc />
    public async Task<Result<List<RecentActivityDto>>> GetRecentActivityAsync(int count = 10, CancellationToken ct = default)
    {
        var rows = await _db.ProjectActivities
            .AsNoTracking()
            .Include(a => a.User)
            .Include(a => a.Project)
            .OrderByDescending(a => a.CreatedAt)
            .Take(count)
            .ToListAsync(ct);

        var activities = rows.Select(a => new RecentActivityDto(
            a.Id,
            a.Action,
            a.Detail,
            a.CreatedAt,
            a.User != null ? $"{a.User.FirstName} {a.User.LastName}" : null,
            a.Project?.Name
        )).ToList();

        return Result<List<RecentActivityDto>>.Ok(activities);
    }

    /// <inheritdoc />
    public async Task<Result<(int Count, List<ProjectSummaryDto> Projects)>> GetOverdueProjectsAsync(CancellationToken ct = default)
    {
        var now = DateTime.UtcNow;
        var query = _db.Projects
            .AsNoTracking()
            .Where(p => p.EndDate < now && p.Status != ProjectStatus.Completed && p.Status != ProjectStatus.Archived)
            .Include(p => p.ProjectManager)
            .OrderByDescending(p => p.EndDate);

        var count = await query.CountAsync(ct);
        var rawProjects = await query.ToListAsync(ct);

        var projects = rawProjects.Select(p => new ProjectSummaryDto(
            p.Id,
            p.Name,
            p.ClientName,
            p.ProjectManager != null ? $"{p.ProjectManager.FirstName} {p.ProjectManager.LastName}" : null,
            p.ProjectManager?.Email,
            p.Status,
            p.StartDate,
            p.EndDate,
            p.CreatedAt
        )).ToList();

        return Result<(int, List<ProjectSummaryDto>)>.Ok((count, projects));
    }
}

File: web/back-nest/src/dashboard/dashboard.service.ts
import { Injectable } from '@nestjs/common';
import { PrismaService } from '../prisma/prisma.service.js';
import { Result } from '../common/result.js';

@Injectable()
export class DashboardService {
  constructor(private readonly prisma: PrismaService) {}

  async getStats() {
    const now = new Date();
    const startOfMonth = new Date(now.getFullYear(), now.getMonth(), 1);

    const [total, active, completed, archived, overdue, thisMonth] = await Promise.all([
      this.prisma.project.count({ where: { isDeleted: false } }),
      this.prisma.project.count({ where: { isDeleted: false, status: { in: ['Draft', 'Kickoff', 'Realisation'] } } }),
      this.prisma.project.count({ where: { isDeleted: false, status: 'Completed' } }),
      this.prisma.project.count({ where: { isDeleted: false, status: 'Archived' } }),
      this.prisma.project.count({ where: { isDeleted: false, endDate: { lt: now }, status: { notIn: ['Cloture', 'Completed', 'Archived'] } } }),
      this.prisma.project.count({ where: { isDeleted: false, createdAt: { gte: startOfMonth } } }),
    ]);

    const byStatus = await this.getProjectsByStatus();
    const byPriority = await this.prisma.project.groupBy({ by: ['priority'], where: { isDeleted: false }, _count: true });
    const workloads = await this.getWorkloads();
    const recentActivities = await this.getRecentActivity(10);

    return Result.ok({
      totalProjects: total,
      activeProjects: active,
      completedProjects: completed,
      archivedProjects: archived,
      overdueProjects: overdue,
      projectsThisMonth: thisMonth,
      projectsByStatus: byStatus.value,
      projectsByPriority: Object.fromEntries(byPriority.map((g) => [g.priority, g._count])),
      projectManagerWorkloads: workloads.value,
      recentActivities: recentActivities.value,
    });
  }

  async getProjectsByStatus() {
    const groups = await this.prisma.project.groupBy({ by: ['status'], where: { isDeleted: false }, _count: true });
    return Result.ok(Object.fromEntries(groups.map((g) => [g.status, g._count])));
  }

  async getWorkloads() {
    const managers = await this.prisma.appUser.findMany({
      where: { role: 'ProjectManager', isActive: true },
      include: {
        managedProjects: { where: { isDeleted: false }, select: { status: true, endDate: true } },
      },
    });

    return Result.ok(
      managers.map((m) => ({
        managerId: m.id,
        managerName: `${m.firstName} ${m.lastName}`,
        managerEmail: m.email,
        totalProjects: m.managedProjects.length,
        inProgressProjects: m.managedProjects.filter((p) => ['Kickoff', 'Realisation'].includes(p.status)).length,
        completedProjects: m.managedProjects.filter((p) => p.status === 'Completed').length,
        overdueProjects: m.managedProjects.filter((p) => p.endDate < new Date() && !['Cloture', 'Completed', 'Archived'].includes(p.status)).length,
      })),
    );
  }

  async getRecentActivity(count = 10) {
    // Cap to [1, 500] and reject NaN/negative inputs.
    const safeCount = Number.isFinite(count) && count > 0 ? Math.min(Math.floor(count), 500) : 10;
    const activities = await this.prisma.projectActivity.findMany({
      include: {
        project: { select: { id: true, name: true, clientName: true } },
        user: { select: { id: true, firstName: true, lastName: true, role: true } },
      },
      orderBy: { createdAt: 'desc' },
      take: safeCount,
    });

    return Result.ok(
      activities.map((a) => ({
        id: a.id,
        action: a.action,
        detail: a.detail,
        timestamp: a.createdAt,
        userName: a.user ? `${a.user.firstName} ${a.user.lastName}` : null,
        userId: a.user?.id ?? null,
        userRole: a.user?.role ?? null,
        projectId: a.project?.id ?? null,
        projectName: a.project?.name ?? null,
        projectClientName: a.project?.clientName ?? null,
      })),
    );
  }

  async getActivityStats() {
    const now = new Date();
    const startOfToday = new Date(now.getFullYear(), now.getMonth(), now.getDate());
    const startOfWeek = new Date(startOfToday);
    startOfWeek.setDate(startOfToday.getDate() - startOfToday.getDay());

    const [totalToday, totalThisWeek, projectCounts] = await Promise.all([
      this.prisma.projectActivity.count({ where: { createdAt: { gte: startOfToday } } }),
      this.prisma.projectActivity.count({ where: { createdAt: { gte: startOfWeek } } }),
      this.prisma.projectActivity.groupBy({
        by: ['projectId'],
        _count: { _all: true },
        orderBy: { _count: { projectId: 'desc' } },
        take: 1,
      }),
    ]);

    let mostActiveProject: { id: string; name: string; count: number } | null = null;
    if (projectCounts.length > 0) {
      const topEntry = projectCounts[0];
      const project = await this.prisma.project.findUnique({
        where: { id: topEntry.projectId },
        select: { id: true, name: true },
      });
      if (project) {
        mostActiveProject = { id: project.id, name: project.name, count: topEntry._count._all };
      }
    }

    return Result.ok({ totalToday, totalThisWeek, mostActiveProject });
  }

  async getOverdueProjects() {
    const projects = await this.prisma.project.findMany({
      where: { isDeleted: false, endDate: { lt: new Date() }, status: { notIn: ['Cloture', 'Completed', 'Archived'] } },
      include: { projectManager: { select: { firstName: true, lastName: true, email: true } } },
    });
    return Result.ok({ count: projects.length, projects });
  }
}
